PyInstaller/utils/hooks/__init__.py

```diff
@@ -729,6 +729,13 @@ def collect_data_files(package, include_py_files=False, subdir=None,
     if not include_py_files:
         excludes += ['**/*' + s for s in ALL_SUFFIXES]
 
+    # Exclude __pycache__ directories if include_py_files is True.
+    if include_py_files:
+        excludes.append('**/__pycache__')
+
+    # Exclude .pyo files whether include_py_files is True or False.
+    excludes.append('**/*.pyo')
```

Comment:
Exclude only if they are not in compat.ALL_SUFFIXES. See https://github.com/pyinstaller/pyinstaller/issues/5115#issuecomment-686970098 for explanasion

More precisly
```suggestion
    if ".pyo" not in ALL_SUFFIXES:
        excludes.append('**/*.pyo')
    if ".pyc" not in ALL_SUFFIXES:
        excludes.append('**/*.pyc')
```

Reply:
RE 
```    if ".pyc" not in ALL_SUFFIXES:
        excludes.append('**/*.pyc')
```

Hi @htgoebel , do we need to append `.pyc` as per your suggestion when `.pyc`  is always in [ALL_SUFFIXES](https://github.com/pyinstaller/pyinstaller/blob/master/PyInstaller/utils/hooks/__init__.py#L730) by default? Or am I missing something? 

It is assigned to the BYTECODE_SUFFIXES list that is a part of the ALL_SUFFIXES variable exported by CPython [here](https://github.com/python/cpython/blob/3.5/Lib/importlib/_bootstrap_external.py#L272) and ultimately imported into the ALL_SUFFIXES variable in PyInstaller.

Reply:
Hi @htgoebel. regarding 
```
    if ".pyc" not in ALL_SUFFIXES:
        excludes.append('**/*.pyc')
```
...are you sure we need to append `'**/*.pyc'`? It is *always* n `ALL_SUFFIXES` because CPython exports `.pyc` as an element inside the ALL_SUFFIXES list.

See [here](https://github.com/python/cpython/blob/3.5/Lib/importlib/_bootstrap_external.py#L272) for the variable `BYTECODE_SUFFIXES' ( which is returned by CPython's all_suffixes() [here](https://github.com/python/cpython/blob/3.5/Lib/importlib/machinery.py#L21) and then instantiated into PyInstaller [here](https://github.com/pyinstaller/pyinstaller/blob/develop/PyInstaller/compat.py#L169)

Reply:
You are right.

BTW: `.pyo` is no longer in ALL_SUFFIXES since 3.5 https://github.com/python/cpython/blob/3.5/Lib/importlib/_bootstrap_external.py#L272-L274 (see 3.4: https://github.com/python/cpython/blob/3.4/Lib/importlib/_bootstrap.py#L435)

Reply:
No because `excludes.append('**/*.pyo')` is adding directly to `excludes`. If it were `ALL_SUFFIXES.add('.pyo')` then this would be the case. Is that what's causing the confusion?

Reply:
@htgoebel Can we clarify what the intended behaviour is for `.pyc` files with or without `include_py_files`? My understanding is that they should be excluded in either case. But the current code and your comments imply otherwise...

Reply:
> No because `excludes.append('**/*.pyo')` is adding directly to `excludes`. If it were `ALL_SUFFIXES.add('.pyo')` then this would be the case. Is that what's causing the confusion?

You're right - thanks for clarifying that.

Reply:
> No because `excludes.append('**/*.pyo')` is adding directly to `excludes`. If it were `ALL_SUFFIXES.add('.pyo')` then this would be the case. Is that what's causing the confusion?

@bwoodsend  - `ALL_SUFFIXES` is only appended to the `excludes` array if `include_py_files` is false (see [here](https://github.com/pyinstaller/pyinstaller/blob/master/PyInstaller/utils/hooks/__init__.py#L729)). As a result, this code:

```
    if ".pyc" not in ALL_SUFFIXES:
        excludes.append('**/*.pyc')
``` 

... will result in a situation whereby `**/*.pyc` is in `ALL_SUFFIXES`, and therefore not appended to `excludes` - but with `ALL_SUFFIXES` not being appended to `excludes` either, because `include_py_files` is False (again, see [here](https://github.com/pyinstaller/pyinstaller/blob/master/PyInstaller/utils/hooks/__init__.py#L729)). 

As a result, `**/*.pyc` will be collected if `include_py_files` is false.

Reply:
The original intended behavior of `collect_data_files` is to collect data files inside a package. However, in some cases, these data files might need to include source code. In my case, I was packaging an application that looking for `*.py` files in a `plugins` subdirectory; putting these in the archive as hidden imports didn't help, since the Python source expected to find actual `.py` files then import them. Hence, the `include_py_files` option was designed to include those files.

Now, the fun part: should `.pyc/.pyo/_pycache/etc` be included as well? It heavily depends on the intend of the hook and the needs of the application. Here's the docstring snippet:

```
    -   By default, all Python executable files (those ending in ``.py``,
        ``.pyc``, and so on) will NOT be collected; setting the
        ``include_py_files`` argument to ``True`` collects these files as well.
        This is typically used with Python routines (such as those in
        ``pkgutil``) that search a given directory for Python executable files
        then load them as extensions or plugins.
```

I later added both `includes` and `excludes` to provide more control over this. It's gotten to be a bit complex, but the intention of the function is to make it easier for hook authors to include data files (which may include Python source) relative to a module.

Reply:
Well in that case I think the cleanest approach would be to let `.pyc`s et all be collected with `include_py_files=True`. If the hook author knows that they don't want them then they can use `excludes=["**/*.pyc"]`. I think trying to exclude them by default so that the author has to explicitly add them to `includes` could lead to some ugly logic where patterns are in both `includes` and `excludes` and we have to decide which gets priority.